src/operations/execute_operation.ts

```diff
@@ -66,45 +62,46 @@ export interface ExecutionResult {
 export function executeOperation<
   T extends AbstractOperation<TResult>,
   TResult = ResultTypeFromOperation<T>
->(topologyProvider: TopologyProvider, operation: T): Promise<TResult>;
+>(client: MongoClient, operation: T): Promise<TResult>;
 export function executeOperation<
   T extends AbstractOperation<TResult>,
   TResult = ResultTypeFromOperation<T>
->(topologyProvider: TopologyProvider, operation: T, callback: Callback<TResult>): void;
+>(client: MongoClient, operation: T, callback: Callback<TResult>): void;
 export function executeOperation<
   T extends AbstractOperation<TResult>,
   TResult = ResultTypeFromOperation<T>
->(
-  topologyProvider: TopologyProvider,
-  operation: T,
-  callback?: Callback<TResult>
-): Promise<TResult> | void;
+>(client: MongoClient, operation: T, callback?: Callback<TResult>): Promise<TResult> | void;
 export function executeOperation<
   T extends AbstractOperation<TResult>,
   TResult = ResultTypeFromOperation<T>
->(
-  topologyProvider: TopologyProvider,
-  operation: T,
-  callback?: Callback<TResult>
-): Promise<TResult> | void {
+>(client: MongoClient, operation: T, callback?: Callback<TResult>): Promise<TResult> | void {
   if (!(operation instanceof AbstractOperation)) {
     // TODO(NODE-3483): Extend MongoRuntimeError
     throw new MongoRuntimeError('This method requires a valid operation instance');
   }
 
   return maybePromise(callback, callback => {
-    let topology: Topology;
-    try {
-      // TODO(NODE-4151): Use skipPingOnConnect and call connect here to make client.connect optional
-      topology = getTopology(topologyProvider);
-    } catch (error) {
-      return callback(error);
+    const topology = client.topology;
+
+    if (topology == null) {
+      if (client.s.hasBeenClosed) {
+        return callback(new MongoNotConnectedError('client was closed'));
```

Comment:
Can we add a more descriptive error message?  Something like:

"attempted to execute operation with closed client"

Reply:
In the spirit of phrasing things as "must be" I went with this:
`Client must be connected before running operations`